### PR TITLE
Refactor bet structure

### PR DESCRIPTION
**Pull request by kenzal**
Major Refactor of Bet structuring.

- Separated Bets into generalized (craps.bet) and table/actualized (craps.table.bet) all under a BetInterface.
- Bets are now equivalent so long as type (and, where applicable, placement) is the same.  The wager/odds does not figure into equivalence.
- Lists of bets have been replaced with sets taking advantage of the new bet equivalence.
- Union type hinting has largely been done away with in favor of aliased union types for readability
- craps.table.Table was extracted to its own file to prevent circular dependency
- tests were updated to handle the above

### CODE REVIEW

File: craps/table/table.py [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+                if not isinstance(existing_bet, ToggleableBetAbstract):
+                    raise BadBetActionException
+                if bet.same_type_and_place(existing_bet):
+                    existing_bet.turn_on()
+
+    def _process_turn_off(self, bets: list[BetSignature] = None):
+        bets = [BetAbstract.from_signature(signature=signature, table=self) for signature in bets]
+        for bet in bets:
+            for existing_bet in self.bets:
+                if not isinstance(existing_bet, BetAbstract):
```
Comment (jeffreytyler) [inline craps/table/table.py]: it looks like lines 180 to 185 takes a signature and returns a bet. It appears to be used several times through out. Could you move it into a method and reuse it?

Comment (jeffreytyler) [reply craps/table/table.py]: That may be expandable to line 187 with the ToggleableBetAbstract as well.

File: craps/table/bet.py [hunk 2]
```diff
@@ -1,133 +1,72 @@
 """
```
Comment (jeffreytyler) [inline craps/table/bet.py]: This file is still pretty large. Is it worth moving some of the classes our and creating a package for them?

File: craps/table/table.py (hunk 1 above)
Comment (kenzal) [reply craps/table/table.py]: Sounds good, the logic could probably be reorganized to make it cleaner. I'll play around with some truth tables to see what I can make of it.

File: craps/table/bet.py (hunk 2 above)
Comment (kenzal) [reply craps/table/bet.py]: I split the file into two separate files, one for concrete bets one for the abstracts and decorator

File: craps/table/table.py (hunk 1 above)
Comment (kenzal) [reply craps/table/table.py]: It was redundant code anyway, I've removed it from all but the initial load.

File: craps/table/bets.py [hunk 3]
```diff
@@ -0,0 +1,480 @@
...
+    def get_payout(self, outcome: DiceOutcome) -> int:
+        if not self.is_winner(outcome):
+            return 0
+        return int(self.wager / self.multi_bet * (
+            self._table.config.hop_hard_pay_to_one if outcome.is_hard()
+            else self._table.config.hop_easy_pay_to_one))
+
+
+@ignore_placement_for_compare
+class CE(PropBetAbstract):
```
Comment (gorrilla10101) [inline craps/table/bets.py:468]: CE is not a very descriptive name, what about
```suggestion
class CrapsOrEleven(PropBetAbstract):
```

Comment (kenzal) [reply craps/table/bets.py:468]: The actual bet is called the C&E bet

Comment (gorrilla10101) [reply craps/table/bets.py:468]: In that case, it's just a terrible name all together but it at least makes more sense now.

**Review by gorrilla10101 — APPROVED** [review]

**Pull request merged**